File: 739-daily-temperatures/739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        ans, hottest = [0] * len(temperatures), 0
        for i in reversed(range(len(temperatures))):
            t = temperatures[i]
            if hottest <= t:
                hottest = t
                ans[i] = 0
            else:
                j = i + 1
                while j < len(temperatures):
                    if temperatures[j] > t:
                        ans[i] = j - i
                        break
                    else:
                        j += ans[j]
        return ans

File: 739-daily-temperatures/739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] ans = new int[temperatures.length];
        Stack<Integer> stack = new Stack<>();
        for(int i = temperatures.length - 1; i >= 0; i--){
            while(stack.size() > 0 && temperatures[stack.peek()] <= temperatures[i]){
                stack.pop();
            }
            
            if(stack.size() != 0){
                ans[i] = stack.peek() - i;
            }
            stack.add(i);
        }
        return ans;
    }
}